hc_dpki/src/keypair.rs

```diff
@@ -0,0 +1,562 @@
+        offset += 1;
+
+        key_buf.write(1, kk.get_sig())?;
+        offset += sign::PUBLICKEYBYTES;
+
+        key_buf.write(offset, kk.get_enc())?;
+        offset += kx::PUBLICKEYBYTES;
+
+        key_buf.write(offset, &**self.sign_priv.read_lock())?;
+        offset += sign::SECRETKEYBYTES;
+
+        key_buf.write(offset, &**self.enc_priv.read_lock())?;
+
+        let pw_enc: bundle::ReturnBundleData = util::pw_enc(&mut key_buf, passphrase)?;
+        let bundle_data_serialized = json::encode(&pw_enc).unwrap();
+
+        // conver to base64
+        let bundle_data_encoded = base64::encode(&bundle_data_serialized);
+
+        Ok(bundle::KeyBundle {
+            bundle_type,
+            hint,
+            data: bundle_data_encoded,
+        })
+    }
+
+    /// initialize the pairs from an encrypted persistence bundle
+    ///
+    /// @param {object} bundle - persistence info
+    ///
+    /// @param {SecBuf} passphrase - decryption passphrase
+    pub fn from_bundle(
+        bundle: &bundle::KeyBundle,
+        passphrase: &mut SecBuf,
+    ) -> Result<Keypair, HolochainError> {
+        // decoding the bundle.data of type util::ReturnBundledata
+        let bundle_decoded = base64::decode(&bundle.data)?;
+        let bundle_string = str::from_utf8(&bundle_decoded).unwrap();
+        let data: bundle::ReturnBundleData = json::decode(&bundle_string).unwrap();
+        let mut decrypted_data = SecBuf::with_secure(BUNDLE_DATA_LEN);
+        util::pw_dec(&data, passphrase, &mut decrypted_data)?;
+        let mut sign_priv = SecBuf::with_secure(64);
+        let mut enc_priv = SecBuf::with_secure(32);
+
+        let pub_keys = {
+            let decrypted_data = decrypted_data.read_lock();
+            if decrypted_data[0] != 1 {
+                return Err(HolochainError::ErrorGeneric(format!(
+                    "Invalid Bundle Version : {:?}",
+                    decrypted_data[0]
+                )));
+            }
+            sign_priv.write(0, &decrypted_data[65..129])?;
+            enc_priv.write(0, &decrypted_data[129..161])?;
+
+            KeyBuffer::with_raw_parts(
+                array_ref![&decrypted_data, 1, 32],
+                array_ref![&decrypted_data, 33, 32],
+            )
+            .render()
+        };
+
+        Ok(Keypair {
+            pub_keys,
+            enc_priv,
+            sign_priv,
+        })
+    }
+
+    /// sign some arbitrary data with the signing private key
+    ///
+    /// @param {SecBuf} data - the data to sign
+    ///
+    /// @param {SecBuf} signature - Empty Buf the sign
+    pub fn sign(
+        &mut self,
+        data: &mut SecBuf,
+        signature: &mut SecBuf,
+    ) -> Result<(), HolochainError> {
+        sign::sign(data, &mut self.sign_priv, signature)?;
+        Ok(())
+    }
+
+    /// verify data that was signed with our private signing key
+    ///
+    /// @param {SecBuf} signature
+    ///
+    /// @param {SecBuf} data
+    pub fn verify(
```

Comment:
`verify` being a method of `Keypair` means I need the full keypair, including the private key, to verify a signature, no? I don't see a way to construct a `Keypair` without the private key. But we need to verify signatures on other nodes that don't have the private key...

Reply:
`self` is only needed for `pub_keys`. Can we remove self from the signature and instead pass in the public key?

Reply:
yep! we absolutely need a verify that is separate from keypair. I was actually imagining that being a function attached to [KeyBuffer](https://github.com/holochain/holochain-rust/blob/develop/core_types/src/agent.rs#L22) or AgentId, since that's where we'll use it from most often. Was thinking that would be a follow-on though. Can't verify anything until we can sign things : )